### PR TITLE
MWUASSIGNT-21/MWUASSIGNT-22: Two new tasks

### PR DESCRIPTION
**Pull request by omarbenhaddi**
request to a unit
agent

### CODE REVIEW

File: src/java/fr/paris/lutece/plugins/workflow/modules/userassignment/service/task/TaskUserAssignmentNotification.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+		for( AdminUser adminUser :  _userResourceService.listActiveUserByResource( resourceHistory.getIdResource( ), resourceHistory.getResourceType( ) ) )
+		{	
+			Map<String, Object> model = new HashMap< >( );
+			
+			model.put( MARK_RESOURCE_ID, resourceHistory.getIdResource( ) );
+			model.put( MARK_RESOURCE_TYPE, resourceHistory.getResourceType( ) );
+			
+			HtmlTemplate html = AppTemplateService.getTemplateFromStringFtl( config.getMessage( ), locale, model );
+			
+			MailService.sendMailHtml( adminUser.getEmail( ), userConnected.getFirstName( ) , userConnected.getEmail( ), config.getSubject( ), html.getHtml( ) );
```
Comment (seboo) [inline src/java/fr/paris/lutece/plugins/workflow/modules/userassignment/service/task/TaskUserAssignmentNotification.java:148]: the "from" could come from task config

File: src/java/fr/paris/lutece/plugins/workflow/modules/userassignment/service/task/TaskUserAssignmentNotification.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+
+	@Override
+	public void processTask( int nIdResourceHistory, HttpServletRequest request, Locale locale )
+	{
+		ResourceHistory resourceHistory = _resourceHistoryService.findByPrimaryKey( nIdResourceHistory );
+		
+		TaskUserAssignmentNotificationConfig config = _taskConfigService.findByPrimaryKey( getId( ) );
+		
+		if ( resourceHistory != null && config != null )
```
Comment (seboo) [inline src/java/fr/paris/lutece/plugins/workflow/modules/userassignment/service/task/TaskUserAssignmentNotification.java:109]: "resourceHistory" and "config" objects should not be null. I think we don't have to hide this error case : if null, an exception should be throwned.

File: src/java/fr/paris/lutece/plugins/workflow/modules/userassignment/service/task/TaskUserAssignmentNotification.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+		{			
+			List<ITask> listTask = _taskService.getListTaskByIdAction( resourceHistory.getAction( ).getId( ), locale );
+			
+			if ( listTask != null )
+			{
+				for ( ITask task : listTask )
+				{
+					if ( task instanceof AssignUserResourceTask )
```
Comment (seboo) [inline src/java/fr/paris/lutece/plugins/workflow/modules/userassignment/service/task/TaskUserAssignmentNotification.java:117]: We should log if we can't find any instance of AssignUserResourceTask as an error.
This task should also be prior (in display order) to the notification task.

File: src/java/fr/paris/lutece/plugins/workflow/modules/userassignment/service/task/TaskUserAssignmentNotification.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+	 */
+	private void notifyUser( HttpServletRequest request, Locale locale, ResourceHistory resourceHistory, TaskUserAssignmentNotificationConfig config )
+	{
+		AdminUser userConnected = AdminUserService.getAdminUser( request );
+
+		for( AdminUser adminUser :  _userResourceService.listActiveUserByResource( resourceHistory.getIdResource( ), resourceHistory.getResourceType( ) ) )
+		{	
+			Map<String, Object> model = new HashMap< >( );
+			
+			model.put( MARK_RESOURCE_ID, resourceHistory.getIdResource( ) );
```
Comment (seboo) [inline src/java/fr/paris/lutece/plugins/workflow/modules/userassignment/service/task/TaskUserAssignmentNotification.java:143]: What's the use for ?

File: src/java/fr/paris/lutece/plugins/workflow/modules/userassignment/service/task/TaskUserAssignmentNotification.java (hunk 3 above)
Comment (seboo) [reply src/java/fr/paris/lutece/plugins/workflow/modules/userassignment/service/task/TaskUserAssignmentNotification.java:117]: Maybe the AssignUserResourceTask task list of the action could be listed when creating the notification task ?

File: src/java/fr/paris/lutece/plugins/workflow/modules/userassignment/web/task/UserAssignmentNotificationTaskComponent.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    private ITaskService _taskService;
+    
+	@Override
+	public String getDisplayConfigForm( HttpServletRequest request, Locale locale, ITask task )
+	{
+		Map<String, Object> model = new HashMap< >( );
+		
+		TaskUserAssignmentNotificationConfig config = _taskConfigService.findByPrimaryKey( task.getId( ) );
+
+		model.put ( MARK_ASSIGN_RESPONSE_TASK, assignResponseTaskExist( locale, task ) );
```
Comment (seboo) [inline src/java/fr/paris/lutece/plugins/workflow/modules/userassignment/web/task/UserAssignmentNotificationTaskComponent.java:89]: We could store the Id of this linked task.

File: src/sql/plugins/workflow/modules/userassignment/plugin/create_db_userassignment.sql [hunk 6]
```diff
@@ -11,4 +11,15 @@ CREATE TABLE workflow_task_assign_user_information (
...
   information_value VARCHAR(255) NULL
-);
\ No newline at end of file
+);
+
+/*==================================================================*/
+/* Table structure for table workflow_task_userassignment_assignment_notification_cf */
+/*==================================================================*/
+DROP TABLE IF EXISTS workflow_task_userassignment_assignment_notification_cf;
+CREATE TABLE workflow_task_userassignment_assignment_notification_cf (
```
Comment (seboo) [inline src/sql/plugins/workflow/modules/userassignment/plugin/create_db_userassignment.sql:20]: As said before, we could store a specific "From" and the id of the assignment task

**Review by seboo — COMMENTED** [review]
hi
the file [userassignment_messages_en.properties] should be named [userassignment_messages.properties]  (it's the default)

In a general view, the only difference of this task and a generic task that sends a message to the connected user, is that the message is sent only if another task of type "userassignment" is present in the same action... 
This specific task should be able to look for assignment informations of the other task ...
It could be a generic notification task, attached to another task, able to read the workflow history ?

File: src/java/fr/paris/lutece/plugins/workflow/modules/userassignment/service/task/UnassignUserResourceAutomaticTask.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        _assignUserResourceTaskService = assignUserResourceTaskService;
+        _resourceHistoryService = resourceHistoryService;
+    }
+
+    @Override
+    public void processTask( int nIdResourceHistory, HttpServletRequest request, Locale locale )
+    {
+        ResourceHistory resourceHistory = _resourceHistoryService.findByPrimaryKey( nIdResourceHistory );
+
+        if ( resourceHistory != null )
```
Comment (seboo) [inline src/java/fr/paris/lutece/plugins/workflow/modules/userassignment/service/task/UnassignUserResourceAutomaticTask.java:76]: resourceHistory should never be null

File: src/java/fr/paris/lutece/plugins/workflow/modules/userassignment/service/task/TaskUserAssignmentNotification.java (hunk 4 above)
Comment (omarbenhaddi) [reply src/java/fr/paris/lutece/plugins/workflow/modules/userassignment/service/task/TaskUserAssignmentNotification.java:143]: These marks are to be used in the content of the notification message. They will allow us to create URLs for example.

**Pull request merged**